### ISSUE TITLE
False positive in `unit_expr` lint with diverging expressions

### ISSUE THREAD

**Issue by goffrie** [rust-lang/rust-clippy#2095]
This program:
```rust
fn main() {
    let wot = match Ok(1) {
        Ok(x) => x,
        Err(()) => { panic!(); },
    };
    println!("{}", wot);
}
```
gives this spurious error:
```
warning: This expression evaluates to the Unit type ()
 --> src/main.rs:2:15
  |
2 |       let wot = match Ok(1) {
  |  _______________^
3 | |         Ok(x) => x,
4 | |         Err(()) => { panic!(); },
5 | |     };
  | |_____^
  |
  = note: #[warn(unit_expr)] on by default
note: Consider removing the trailing semicolon
 --> src/main.rs:4:22
  |
4 |         Err(()) => { panic!(); },
  |                      ^^^^^^^^^
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.164/index.html#unit_expr
```
[playground link](https://play.rust-lang.org/?gist=cb20cbfabca646bf81a1c2467d1e1bee&version=stable)

Removing the trailing semicolon does silence the warning, but it shouldn't be necessary.

**Comment by birkenfeld** [comment]
Interesting. I just stumbled across this too, and although I wrote the code that triggered the warning, I didn't consciously expect that `diverging();` is considered diverging as well. Although that must be happening at a later stage than the lint is running?

Also, the first span seems to be wrong - the `match` definitely doesn't evaluate to `()`.

**Comment by mockersf** [comment]
No error anymore from the playground, and I couldn't reproduce it.

Do you still have this issue ?

**Comment by oli-obk** [comment]
Just needs a regression test then, I guess

**Comment by Kurt-Bonatz** [comment]
There isn't an issue anymore since the [lint got removed](https://github.com/rust-lang/rust-clippy/blob/075c212849ad2207c0b0ccbd0ec6cc5b1f392275/CHANGELOG.md#00181) in the beginning of 2018. This ticket can probably just be closed.
